llmfoundry/models/llm_embed/modeling_llm_embed.py

```diff
@@ -0,0 +1,447 @@
+            self.infonce_process_group = create_set_process_group(pg_size)
+
+    def construct_model(self):
+        if self.pretrained_model_name_or_path:
+            model = ComposerHFCausalLM.build_inner_model(
+                pretrained=True,
+                pretrained_model_name_or_path=self.
+                pretrained_model_name_or_path,
+                pretrained_lora_id_or_path=self.pretrained_lora_id_or_path,
+                trust_remote_code=self.trust_remote_code,
+                init_device=self.init_device,
+                use_flash_attention_2=self.use_flash_attention_2,
+                use_auth_token=self.use_auth_token,
+                config_overrides=self.config_overrides or {},
+                load_in_8bit=self.load_in_8bit,
+                **self.kwargs,
+            )
+        else:
+            model = MPTForCausalLM(MPTConfig(**self.kwargs))
+            self.is_mpt = True
+        return model
+
+    def format_queries_batch(
+        self,
+        batch: MutableMapping,
+        last_hidden_state: torch.Tensor,
+    ) -> tuple[dict[str, torch.Tensor], torch.Tensor]:
+        """Format `queries` by selecting every ``n``th entry from the batch.
+
+        Here ``n`` is the step size, which represents the number of hard
+        negatives per passage.
+        """
+        queries = {}
+        for key in batch:
+            # Select every `step_size`-th entry from the batch for the given key
+            queries[key] = batch[key][0::self.step_size, :]
+
+        # Select every `step_size`-th entry from `last_hidden_state` along the batch dimension
+        return queries, last_hidden_state[0::self.step_size, :, :]
+
+    def format_passages_batch(
+        self,
+        batch: MutableMapping,
+        last_hidden_state: torch.Tensor,
+    ) -> tuple[dict[str, torch.Tensor], torch.Tensor]:
+        """Format `passages` by selecting every ``n``th entry from the batch.
+
+        Here ``n`` is the step size, which represents the number of hard
+        negatives per passage.
+        """
+        passages = {}
+
+        # Index on a variable step size
+        index = 0
+        for key in batch:
+            num_blocks = batch[key].size(0) // self.step_size
+            index = torch.arange(
+                1,
+                num_blocks * self.step_size + 1,
+                device=last_hidden_state.device,
+            ).view(num_blocks, self.step_size)
+            index = index[:, :self.step_size - 1].reshape(-1)
+            passages[key] = batch[key][index]
+
+        return passages, last_hidden_state[index, :, :]
+
+    def forward(self, batch: MutableMapping) -> CausalLMOutputWithPast:
+        # Collapse pairs into the batch dimension
+        collapse_dims = lambda x: rearrange(x, 'b p d -> (b p) d') if \
+            len(x.shape) > 2 else x
+
+        kwargs = {}
+        for key in batch:
+            kwargs[key] = collapse_dims(batch[key])
```

Comment:
why do we need `kwargs`? Can this not just be `batch[key] = collapse_dims(batch[key])` and then **batch into forward?